design/comment-scanning.md

```diff
@@ -0,0 +1,159 @@
+
+There is a second class of documentation that our analyzer currently doesn’t support. Our linter needs to be able to understand documentation that doesn’t describe already existing features. For example, the user may want to send an instruction to some analyzer-powered tool in its build chain:
+
+```html
+<!-- @registerServiceWorkerHere -->
+```
+
+Or, a user may want to describe some behavior across an entire source range, including any number of other existing features (including zero). For example:
+
+```js
+/* polymer-lint disable: undefined-elements */
+customElements.define('vanilla-element1', AnElementThatActuallyExists);
+customElements.define('vanilla-element2', IPromiseThisExists);
+customElements.define('vanilla-element3', SeriouslyLinterDontBeMad);
+/* polymer-lint enable: undefined-elements */
+```
+
+Instead of building one-off support for this into the analyzer, we’d like to also think about how the analyzer should analyze all comment-based information so that future features don’t need to reinvent the wheel when it comes to how comments should be scanned.
+
+
+## Design Overview
+
+This design groups all analysis-related comments into two supported formats: **Comment Directives** & **Comment Annotations**. **Comment Directives** are standalone features represented by standalone code comments. **Comment Annotations** are additional information/metadata that always describe the existing feature they are "attached" to.
+
+Each supports a different use-case for analysis-related documentation. Together they cover most use cases for passing instructions and metadata to the analyzer in a way that is easy for the user to extend with 3rd party support.
+
+
+### Comment Directive
+
+```
+/* polymer-lint disable */
+/* polymer-lint disable: undefined-elements, rule-2 */
+/* polymer-lint enable: undefined-elements */
+/* lazy-import: '../some-url/some-file.html' */
+<!-- polymer-build:register-service-worker-here -->
+```
+
+- A comment directive represents a single new **Directive** feature type within a document
+- A comment directive may include "arguments", which will be parsed by the scanner
+- A comment directive must exist within its own comment so that the analyzer can properly create a feature that matches that comment node's location
+- Because of the wide range of possible features these could support, specific comment syntax is lax. However they should follow these guidelines:
+  - it is unique enough to never collide with other directive formats
+  - it is formal enough to be matched by substring or RegEx matching
+  - if it takes arguments they should be parsed in the scanning phase
+- Each comment directive will have a scanner responsible for scanning for it.
+  - A scanner may be responsible for scanning multiple directives (ex: `polymer-lint enable`, `polymer-lint disable`)
+Here is a definition for the **Directive** class:
+
+```
+class Directive extends Feature {
```

Comment:
Do we expect this to be subclassed?

Is there any code that will rely on a common interface between Directives? Is this interface needed if not?